src/btreemap.rs

```diff
@@ -575,23 +599,35 @@ where
 
     /// Returns the first key-value pair in the map. The key in this
     /// pair is the minimum key in the map.
-    pub fn first_key_value(&self) -> Option<(K, V)> {
+    pub fn first_key_value(&mut self) -> Option<(K, V)> {
```

Comment:
caching does not have to change the function API. the general approach in rust for caching is to use internal mutability with RefCell<...>.